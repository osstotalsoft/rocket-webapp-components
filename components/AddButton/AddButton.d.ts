import { ButtonProps } from '@material-ui/core/Button';

export interface AddButtonProps extends Omit<ButtonProps, 'color'> {
  /**
  * The title of button.
  */
  title?: string
  /**
  * Callback fired when a "click" event is detected.
  */
  onClick?: () => void
  /**
  * @default "false"
  * If true, button will be disabled.
  */
  disabled?: boolean
  /**
  * @default "theme"
  * The color of button.
  */
  color?: "primary" |
          "info" |
          "theme" |
          "themeNoBackground" |
          "themeWithBackground" |
          "success" |
          "warning" |
          "danger" |
          "rose" |
          "white" |
          "simple" |
          "defaultNoBackground" |
          "primaryNoBackground" |
          "infoNoBackground" |
          "successNoBackground" |
          "warningNoBackground" |
          "dangerNoBackground" |
          "roseNoBackground"
  /**
  *  @default "medium"
  * The size of button. 
  */
  size?: "medium" | "small"
  /**
  *  @default "small"
  * The fontSize of button's content.
  */
  fontSize?: "inherit" | "small" | "medium" | "large"
  /**
  * Override or extend the styles applied to the component
  */
  customClass?: string
  /**
  * @default "false"
  * If true, the keyboard focus ripple will be disabled. 
  */
  disableFocusRipple?: boolean
  /**
  * If true, the ripple effect will be disabled.
  */
  disableRipple?: boolean
  /**
  * @default "false"
  * If given, uses a negative margin to counteract the padding on one side 
  * (this is often helpful for aligning the left or right side of the icon 
  * with content above or below, without ruining the border size and shape).
  */
  edge?: "start" | "end" | "false"
}
/**
 *
 * Demos:
 *
 * - https://bit.dev/totalsoft_oss/react-mui/add-button
 *
 *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/AddButton/AddButton.js
 * 
 * Bit component used:
 * 
 * - IconButton: https://bit.dev/totalsoft_oss/react-mui/icon-button
 * 
 */
export default function AddButton(props: AddButtonProps): JSX.Element;