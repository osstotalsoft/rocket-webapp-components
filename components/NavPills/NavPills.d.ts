import { TabProps } from '@material-ui/core/Tab'
import { TabsProps } from '@material-ui/core/Tabs'

export interface Props {
    /**
     * @default 0 
    * Index of the default active pill
     */
    active?: number;
    /**
     * The content of the component
     */
    tabs: {
        tabButton: string,
        tabIcon: Object,
        tabContent: React.ReactNode
    }[];

    /**
     * @default "theme" 
    * The color of the component
     */
    color?: "primary" |
            "warning" |
            "danger" |
            "success" |
            "info" |
            "rose" |
            "theme"

    /**
     * If set to true, the tabs will be displayed horizontal
     * @default true
     */
    horizontal?: boolean

    /**
    * If set to true, the alignment of pills will be centered
    */
    alignCenter?: boolean;

    /**
    * This property is required
    * @param {object} event The event source of the callback.
    * @param {any} value We default to the index of the child (number)
    */
    onChange: (event: React.SyntheticEvent, value: any) => void;

    /**
    * @default [] 
    * A list of actions available in this component
    */
    actions?: string[];

    /**
     * @default "standard"
     * Determines additional display behavior of the tabs:
        - scrollable will invoke scrolling properties and allow for horizontally scrolling (or swiping) of the tab bar.
        - fullWidth will make the tabs grow to use all the available space, which should be used for small views, like on mobile.
        - standard will render the default state.
     */
    variant?:   "scrollable" |
                "fullWidth" |
                "standard"

    /**
    * Demos:
    * 
    * - https://bit.dev/totalsoft_oss/react-mui/nav-pills
    * 
    *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/NavPills/NavPills.js
    * 
    * Material-UI components used:
    * 
    * - Tab: https://material-ui.com/api/tab/
    * - Tabs: https://material-ui.com/api/tabs/
    * 
    */
}

export type NavPillsProps = Props | TabProps | TabsProps

export default function NavPills(props: NavPillsProps): JSX.Element;