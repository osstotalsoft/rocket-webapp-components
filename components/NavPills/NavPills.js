import React from "react";
import cx from "classnames";
import PropTypes from "prop-types";
import { Tab, Tabs, Grid, Box, makeStyles } from "@material-ui/core";
import navPillsStyle from "./navPillsStyle.js";

const useStyles = makeStyles(navPillsStyle);

const NavPills = ({ tabs, color, horizontal, alignCenter, active, onChange, actions, variant }) => {
    const classes = useStyles();
    const tabButtons = (
        <Box className={classes.container} >
            <Box className={classes.tabsContainer} >
                <Tabs
                    className={classes.root}
                    TabIndicatorProps={{ className: classes.displayNone }}
                    orientation={horizontal ? "horizontal" : "vertical"}
                    value={active}
                    onChange={onChange}
                    centered={alignCenter}
                    variant={variant}
                >
                    {tabs.map((tab, key) => {
                        var icon = {};
                        if (tab.tabIcon !== undefined) {
                            icon["icon"] = <tab.tabIcon className={classes.tabIcon} />;
                        }
                        const pillsClasses =
                            classes.pills +
                            " " +
                            cx({
                                [classes.verticalPills]: !horizontal,
                                [classes.pillsWithIcons]: tab.tabIcon !== undefined
                            });
                        return (
                            <Tab
                                label={tab.tabButton}
                                value={tab.value}
                                key={key}
                                {...icon}
                                classes={{
                                    root: pillsClasses,
                                    selected: classes[color]
                                }}
                            />
                        );
                    })}
                </Tabs>
            </Box>
        </Box>
    );

    const tabActions = (
        <Box className={classes.container} >
            {actions.map(
                (action, index) => (
                    <Box key={index} ml={1}>
                        {action}
                    </Box>
                )
            )}
        </Box>
    )

    const tabContent = (
        <div className={classes.contentWrapper}>
            {tabs.map((tab, key) => {
                return (
                    (key === active) && (<div className={classes.tabContent} key={key}>
                        {tab.tabContent}
                    </div>)
                );
            })}
        </div>
    );

    return horizontal ? (<>
        <Box className={classes.grid} display="flex">
            <Box flexGrow={1} className={classes.tabItemGrid}>{tabButtons}</Box>
            <Box className={classes.itemGrid}>{tabActions}</Box>
        </Box>
        <Box>
            <Box className={classes.itemGrid}>{tabContent}</Box>
        </Box>
    </>) : (<Grid container>
            <Grid item lg={2}>{tabButtons}</Grid>
            <Grid item lg={9}>{tabContent}</Grid>
            <Grid item lg={1}>{tabActions}</Grid>
        </Grid>);
}

NavPills.defaultProps = {
    active: 0,
    color: "primary",
    actions: [],
    horizontal: true,
    variant: "standard"
};

NavPills.propTypes = {
    /**
    * Index of the default active pill
    */
    active: PropTypes.number,
    /**
    * The content of the component
    */
    tabs: PropTypes.arrayOf(
        PropTypes.shape({
            tabButton: PropTypes.string,
            tabIcon: PropTypes.object,
            tabContent: PropTypes.node
        })
    ).isRequired,
    /**
    * The color of the component
    */
    color: PropTypes.oneOf([
        "primary",
        "warning",
        "danger",
        "success",
        "info",
        "rose",
        "theme"
    ]),
    horizontal: PropTypes.bool,
    alignCenter: PropTypes.bool,
    /**
    * This property is required
    * @param {object} event The event source of the callback.
    * @param {any} value We default to the index of the child (number)
    */
    onChange: PropTypes.func.isRequired,
    /**
    * A list of actions available in this component
    */
    actions: PropTypes.array,
    /**
     * Determines additional display behavior of the tabs:
        - scrollable will invoke scrolling properties and allow for horizontally scrolling (or swiping) of the tab bar.
        - fullWidth will make the tabs grow to use all the available space, which should be used for small views, like on mobile.
        - standard will render the default state.
     */
    variant: PropTypes.oneOf([
        "scrollable",
        "fullWidth",
        "standard"
    ]),
};

export default NavPills;