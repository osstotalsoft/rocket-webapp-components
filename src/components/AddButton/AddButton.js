import React from "react";
import PropTypes from "prop-types";
import AddIcon from '@material-ui/icons/Add';
import CustomIconButton from '../IconButton';


const AddButton = ({ title, onClick, disabled, fontSize, ...rest }) => (
    <CustomIconButton aria-label="Add" onClick={onClick} disabled={disabled} tooltip={title} {...rest}>
        <AddIcon fontSize={fontSize} />
    </CustomIconButton>
)

AddButton.defaultProps = {
    color: "theme",
    fontSize: "small",
    size: "medium"
};

AddButton.propTypes = {
    /**
   * The title of button.
   */
    title: PropTypes.string,
    /**
    * Callback fired when a "click" event is detected.
    */
    onClick: PropTypes.func,
    /**
   * If true, the button will be disabled.
   */
    disabled: PropTypes.bool,
    /**
   * The color of button.
   */
    color: PropTypes.oneOf([
        "primary",
        "info",
        "theme",
        "themeNoBackground",
        "themeWithBackground",
        "success",
        "warning",
        "danger",
        "rose",
        "white",
        "simple",
        "defaultNoBackground",
        "primaryNoBackground",
        "infoNoBackground",
        "successNoBackground",
        "warningNoBackground",
        "dangerNoBackground",
        "roseNoBackground",
        "themeWithGradient"
    ]),
    /**
      * The size of button.
      */
    size: PropTypes.oneOf([
        "small",
        "medium"
    ]),
    /**
    * The fontSize of button's content.
    */
    fontSize: PropTypes.oneOf([
        "inherit",
        "small",
        "medium",
        "large"
    ])
};

export default AddButton;