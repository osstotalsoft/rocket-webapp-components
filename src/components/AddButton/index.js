import AddButton from "./AddButton";
export default AddButton;