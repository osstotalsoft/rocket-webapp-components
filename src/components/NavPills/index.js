import NavPills from "./NavPills";
export default NavPills;