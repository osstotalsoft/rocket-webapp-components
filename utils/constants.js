export const emptyArray = []
export const emptyObject = {}
export const emptyString = ''
export const emptyFunction = () => {}
