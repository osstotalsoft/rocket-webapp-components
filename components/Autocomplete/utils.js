import { createFilterOptions } from '@material-ui/lab/Autocomplete'
import { prop, map, innerJoin, find, propEq, all, includes, is, isEmpty, isNil, props, omit } from 'ramda'
import { emptyArray } from '../../utils/constants'

export const findFirstNotNil = (propNames, option) => find(x => !isNil(x), props(propNames, option))
export const isStringOrNumber = option => is(String, option) || is(Number, option)

const hasStringOptions = options => all(is(String), options) && !isEmpty(options)

const filter = createFilterOptions()
export const filterOptions = (labelKey, valueKey, creatable) => (options, params) => {
  const filtered = filter(options, params)
  const { inputValue } = params

  // Suggest the creation of a new value if it's not empty and it doesn't already exist
  if (creatable && inputValue !== '' && !find(propEq(labelKey, inputValue), options)) {
    filtered.push(
      hasStringOptions(options)
        ? {
            _primitiveValue: inputValue,
            _createdOption: true
          }
        : {
            [valueKey]: inputValue,
            [labelKey]: inputValue,
            _createdOption: true
          }
    )
  }

  return filtered
}

export const getSimpleValue = (options, value, valueKey, isMultiSelection) => {
  if (isMultiSelection && (!is(Array, value) || isEmpty(options))) return emptyArray
  if (!all(is(Object), options)) return value

  // Add new options if the Autocomplete is multiSelection and creatable
  if (is(Array, value)) {
    const optionsSimpleValues = map(prop(valueKey), options)
    value?.map(v => {
      if (!includes(v, optionsSimpleValues)) options.push({ [valueKey]: v })
    })
  }

  const result = isMultiSelection ? innerJoin((o, v) => o[valueKey] === v, options, value) : find(propEq(valueKey, value), options)
  return result || null
}

export const computeChangedMultiValue = (input, simpleValue, valueKey, labelKey) =>
  simpleValue
    ? input.map(a => (is(String, a) ? a : findFirstNotNil([valueKey, labelKey, '_primitiveValue'], a)))
    : input.map(a => (is(String, a) ? a : prop('_primitiveValue', a) || omit(['_createdOption'], a)))

export const computeChangedSingleValue = (input, simpleValue, valueKey, labelKey) =>
  simpleValue ? findFirstNotNil([valueKey, labelKey], input) : prop('_primitiveValue', input) ?? omit(['_createdOption'], input)
