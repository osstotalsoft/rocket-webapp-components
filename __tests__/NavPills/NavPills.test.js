import React from "react";
import renderer from "react-test-renderer";
import NavPills from "../../components/NavPills/NavPills";
import AddButton from "../../components/AddButton/AddButton";
import { mount } from "enzyme";
import { Tab } from "@material-ui/core";
import { emptyArray, emptyFunction } from "../../utils/constants";

const tabsArray = [
  {
    tabButton: "Button1",
    tabContent: <div>Content1</div>
  },
  {
    tabButton: "Button2",
    tabContent: <div>Content2</div>
  }
];

describe("NavPills", () => {
  it("Snapshot", () => {
    const component = renderer.create(
      <NavPills
        label="NavPills"
        color="primary"
        active={0}
        tabs={tabsArray}
        onChange={emptyFunction}
        actions={[<AddButton onClick={emptyFunction} />]}
      />
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it("active, color and actions props should be equal to the default values provided", () => {
    const wrapper = mount(
      <NavPills label="NavPills" tabs={tabsArray} onChange={emptyFunction} />
    );

    expect(wrapper.props().color).toBe("primary");
    expect(wrapper.props().actions).toStrictEqual(emptyArray);
    expect(wrapper.props().active).toBe(0);
  });

  it("should have two children", () => {
    const wrapper = mount(
      <NavPills label="NavPills" tabs={tabsArray} onChange={emptyFunction} />
    );

    expect(wrapper.find(Tab)).toHaveLength(2);
  });

  it("Tabs should get the values from the tabs array provided", () => {
    const wrapper = mount(
      <NavPills label="NavPills" tabs={tabsArray} onChange={emptyFunction} />
    );

    expect(
      wrapper
        .find(Tab)
        .at(0)
        .props().label
    ).toBe(tabsArray[0].tabButton);
    expect(
      wrapper
        .find(Tab)
        .at(1)
        .props().label
    ).toBe(tabsArray[1].tabButton);
  });
});
