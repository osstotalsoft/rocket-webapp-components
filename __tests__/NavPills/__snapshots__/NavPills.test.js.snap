// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavPills Snapshot 1`] = `
Array [
  <div
    className="MuiBox-root MuiBox-root-21 makeStyles-grid-11"
  >
    <div
      className="MuiBox-root MuiBox-root-22 makeStyles-tabItemGrid-12"
    >
      <div
        className="MuiBox-root MuiBox-root-23 makeStyles-container-2"
      >
        <div
          className="MuiBox-root MuiBox-root-24 makeStyles-tabsContainer-1"
        >
          <div
            className="MuiTabs-root makeStyles-root-3"
          >
            <div
              className="MuiTabs-scroller MuiTabs-fixed"
              onScroll={[Function]}
              style={
                Object {
                  "marginBottom": null,
                  "overflow": "hidden",
                }
              }
            >
              <div
                className="MuiTabs-flexContainer"
                onKeyDown={[Function]}
                role="tablist"
              >
                <button
                  aria-selected={true}
                  className="MuiButtonBase-root MuiTab-root makeStyles-pills-6  MuiTab-textColorInherit Mui-selected makeStyles-primary-14"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="tab"
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiTab-wrapper"
                  >
                    Button1
                  </span>
                  <span
                    className="PrivateTabIndicator-root-25 PrivateTabIndicator-colorSecondary-27 makeStyles-displayNone-4"
                    style={Object {}}
                  />
                </button>
                <button
                  aria-selected={false}
                  className="MuiButtonBase-root MuiTab-root makeStyles-pills-6  MuiTab-textColorInherit"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  role="tab"
                  tabIndex={-1}
                  type="button"
                >
                  <span
                    className="MuiTab-wrapper"
                  >
                    Button2
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root MuiBox-root-29 makeStyles-itemGrid-13"
    >
      <div
        className="MuiBox-root MuiBox-root-30 makeStyles-container-2"
      >
        <div
          className="MuiBox-root MuiBox-root-31"
        >
          <button
            aria-label="Add"
            className="MuiButtonBase-root MuiIconButton-root makeStyles-button-32 makeStyles-theme-47"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-51"
  >
    <div
      className="MuiBox-root MuiBox-root-52 makeStyles-itemGrid-13"
    >
      <div
        className="makeStyles-contentWrapper-10"
      >
        <div>
          <div>
            Content1
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
